File: 01. Fundamentals/program.py
from abc import ABC, abstractmethod


class User:
    def __init__(self, username, password):
        self.username = username
        self.password = password

    #setters
    def set_username(self, username):
        self.username = username

    def set_password(self, password):
        self.password = password

    #getters
    def get_username(self):
        return self.username
    
    def get_password(self):
        return self.password


class InputDevice(ABC):
    @abstractmethod
    def input(self, data):
        pass


class Keyboard(InputDevice):
    def input(self, data):
        print("---- Inputing data from keyboard ------------------")
        print("Step1: Listen to data from keyboard")
        print("Step2: Pick data from Keyboard")
        print("Step3: Locate current cursor position")
        print("Step4: Place data in the current cursor position")
        print("Step5: Start Listening to data from keyboard")
        print("")  # Empty Print statement at end of every method


class Mouse(InputDevice):
    def input(self, data):
        print("---- Inputing data from mouse ------------------")
        print("Step1: Listen to data from mouse")
        print("Step2: Pick data from mouse")
        print("Step3: Locate current cursor position")
        print("Step4: Place data in the current cursor position")
        print("Step5: Start Listening to data from mouse")
        print("")  # Empty Print statement at end of every method


class TouchScreen(InputDevice):
    def input(self, data):
        print("---- Input Process using Screen Touch ------------------")
        print("Step1: Listen to data from Screen")
        print("Step2: Pick data from Screen")
        print("Step3: Locate current tourch position")
        print("Step4: Place data in the current tourch position")
        print("Step5: Start Listening to data from Screen")
        print("")  # Empty Print statement at end of every method


class ProcessorChip(ABC):
    @abstractmethod
    def process(self, data):
        pass


class Intel(ProcessorChip):
    def process(self, data):
        print("---- Proceessing using Intel Chip ------------------")
        print("")  # Empty Print statement at end of every method


class AMD(ProcessorChip):
    def process(self, data):
        print("---- Proceessing using AMD Chip ------------------")
        print("")  # Empty Print statement at end of every method


class Nvidia(ProcessorChip):
    def process(self, data):
        print("---- Proceessing using Nvidia ------------------")
        print("")  # Empty Print statement at end of every method


class Memory(ABC):
    @abstractmethod
    def store(self, data):
        pass


class SSD(Memory):
    def store(self, data):
        print("--------storing data on SSD----------\n")


class InternalMemory(Memory):
    def store(self, data):
        print("---- Storage Process ------------------")
        print("Step1: Receive data to be stored")
        print("Step2: Open Internal memory where data is to stored")
        print("Step3: Prepare for data storage operation")
        print("Step4: Launch storage operation")
        print("Step5: Send back signal representing the state of the storage operation")
        print("")  # Empty Print statement at end of every method


class OutputDevice(ABC):
    @abstractmethod
    def output(self, data):
        pass


class Monitor(OutputDevice):
    def output(self, data):
        print("---- Outputing to Monitor ------------------")
        print("Step1: Receive data to be output")
        print("Step2: Open monitor where data is to displayed")
        print("Step3: Prepare for data output operation")
        print("Step4: Launch output operation")
        print("Step5: Send back signal representing the state of the output operation")
        print("")


class Projector(OutputDevice):
    def output(self, data):
        print("---- Outputing to Projector ------------------")
        print("Step1: Receive data to be output")
        print("Step2: Open projector where data is to displayed")
        print("Step3: Prepare for data output operation")
        print("Step4: Launch output operation")
        print("Step5: Send back signal representing the state of the output operation")
        print("")



class Computer(ABC):
    def __init__(self, input_device, processor_chip, memory, output_device):
        # Fields
        self.brand = ""
        self.model = ""
        self.user = User(username=None, password=None)
        self.input_device = input_device
        self.processor_chip = processor_chip
        self.memory = memory
        self.output_device = output_device

    #setters
    def set_input(self, input_device:str):
        self.input_device = input_device

    def set_processor_chip(self, processor_chip):
        self.processor_chip = processor_chip

    def set_memory(self, memory):
        self.memory = memory

    def set_output_device(self, output_device):
        self.output_device = output_device   

    #getters
    def get_input(self):
        return self.input_device

    def get_processor_chip(self):
        return self.processor_chip
    
    def get_memory(self):
        return self.memory
    
    def get_output_device(self):
        return self.output_device



    # Methods
    def input(self, data):
        self.input_device.input(data)

    def process(self, data):
        self.processor_chip.process(data)  # delegation of one method to another

    def store(self, data):
        self.memory.store(data)

    def output(self, data):
        self.output_device.output(data)


# Inheritance: Desktop is inheriting from Computer
class Desktop(Computer):
    # Fields
    pass
    # Methods


# Inheritance: Laptop is inheriting from Computer
class Laptop(Computer):
    # Fields
    pass

    # Methods
    def fold(self):
        print("-----------Folding Process ----------")
        print("Step1: Folding")
        print("")


# Inheritance: Walltop is inheriting from Computer
class Walltop(Computer):
    # Fields
    pass



user = User("Shafic", "profic")

print(user.get_username(), user.get_password())

user.set_username("Bruce")
user.set_password("admin")

print(user.get_username(), user.get_password())

computer = Desktop(Keyboard(), Nvidia(), InternalMemory(), Projector())
computer.set_input(TouchScreen())
computer.input("blah")
computer.set_memory(SSD())
computer.process("see")
computer.store("1")
computer.output("12")


computer = Laptop(Mouse(), AMD(), SSD(), Projector())
computer.input("soo")
computer.process("ts")
computer.store("2")
computer.set_output_device(Monitor())
computer.output("12")

computer = Walltop(Keyboard(), Intel(), SSD(), Monitor())
computer.set_processor_chip(Nvidia())
computer.process("s")
#computer.set_memory(InternalMemory())
computer.store("3")
computer.set_input(TouchScreen())
computer.input("simsim")
computer.output("abs")


# my_laptop = Laptop()
# my_laptop.input()
# my_laptop.process(12, "hop")
# my_laptop.store("one")
# my_laptop.output()
# my_laptop.fold()

File: 01. Fundamentals/program.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var User = /** @class */ (function () {
    function User() {
        // Fields
        this.username = 'johndoe';
        this.password = 'pswd';
    }
    return User;
}());
var KeyBoard = /** @class */ (function () {
    function KeyBoard() {
    }
    KeyBoard.prototype.input = function (data) {
        console.log("---- Data input from Keyboard -------------------");
        console.log("Step 1: Listen to data from keyboard");
        console.log("Step 2: Pick data from keyboard");
        console.log("Step 3: Locate current cursor position");
        console.log("Step 4: Place data to the current cursor position");
        console.log("Step 5: Start listening to data from keyboard");
        console.log("");
    };
    return KeyBoard;
}());
var Mouse = /** @class */ (function () {
    function Mouse() {
    }
    Mouse.prototype.input = function (data) {
        console.log("---- Data input from Mouse -------------------");
        console.log("Step 1: Listen to data from Mouse");
        console.log("");
    };
    return Mouse;
}());
var Intel = /** @class */ (function () {
    function Intel() {
    }
    Intel.prototype.process = function (data) {
        console.log("--------------------");
        console.log("Processing using Intel Chip..");
        console.log("");
    };
    return Intel;
}());
var AMD = /** @class */ (function () {
    function AMD() {
    }
    AMD.prototype.process = function (data) {
        console.log("--------------------");
        console.log("Processing using AMD Chip..");
        console.log("");
    };
    return AMD;
}());
var Nvidia = /** @class */ (function () {
    function Nvidia() {
    }
    Nvidia.prototype.process = function (data) {
        console.log("--------------------");
        console.log("Processing using Invidia Chip..");
        console.log("");
    };
    return Nvidia;
}());
var InternalMemory = /** @class */ (function () {
    function InternalMemory() {
    }
    InternalMemory.prototype.store = function (data) {
        console.log("---- Storing data in internal memory -------------------");
        console.log("Step 1: Receive data to be stored");
        console.log("Step 2: Open internal memory where data is to be stored");
        console.log("Step 3: Prepare for data storage");
        console.log("Step 4: launch storage operation");
        console.log("Step 5: Send back signal representing the state of the storage operation");
        console.log("");
    };
    return InternalMemory;
}());
var SSD = /** @class */ (function () {
    function SSD() {
    }
    SSD.prototype.store = function (data) {
        console.log("storing data on SSD");
    };
    return SSD;
}());
var HDD = /** @class */ (function () {
    function HDD() {
    }
    HDD.prototype.store = function (data) {
        console.log("storing data on HDD");
    };
    return HDD;
}());
var Monitor = /** @class */ (function () {
    function Monitor() {
    }
    Monitor.prototype.output = function (data) {
        console.log("---- Data output from Monitor -------------------");
        console.log("");
    };
    return Monitor;
}());
var Projector = /** @class */ (function () {
    function Projector() {
    }
    Projector.prototype.output = function (data) {
        console.log("---- Data output from Projector -------------------");
        console.log("");
    };
    return Projector;
}());
// Computer Classes
var Computer = /** @class */ (function () {
    // Methods
    function Computer(brand, model, inputDevice, proccessorChip, storageDevice, outputDevice) {
        this.user = new User();
        this.brand = brand;
        this.model = model;
        this.inputDevice = inputDevice;
        this.proccessorChip = proccessorChip;
        this.storageDevice = storageDevice;
        this.outputDevice = outputDevice;
    }
    Computer.prototype.input = function (data) {
        this.inputDevice.input(data);
    };
    Computer.prototype.process = function (data) {
        this.proccessorChip.process(data);
    };
    Computer.prototype.store = function (data) {
        this.storageDevice.store(data);
    };
    Computer.prototype.output = function (data) {
        this.outputDevice.output(data);
    };
    // Setters
    Computer.prototype.setBrand = function (brand) {
        this.brand = brand;
    };
    Computer.prototype.setModel = function (model) {
        this.model = model;
    };
    Computer.prototype.setInputDevice = function (inputDevice) {
        this.inputDevice = inputDevice;
    };
    Computer.prototype.setProcessorChip = function (proccessorChip) {
        this.proccessorChip = proccessorChip;
    };
    Computer.prototype.setStorageDevice = function (storageDevice) {
        this.storageDevice = storageDevice;
    };
    Computer.prototype.setOutputDevice = function (outputDevice) {
        this.outputDevice = outputDevice;
    };
    //getters
    Computer.prototype.getBrand = function () {
        return this.brand;
    };
    Computer.prototype.getModel = function () {
        return this.model;
    };
    Computer.prototype.getInputDevice = function () {
        return this.inputDevice;
    };
    Computer.prototype.getProcessorChip = function () {
        return this.proccessorChip;
    };
    Computer.prototype.getStorageDevice = function () {
        return this.storageDevice;
    };
    Computer.prototype.getOutputDevice = function () {
        return this.outputDevice;
    };
    return Computer;
}());
var Desktop = /** @class */ (function (_super) {
    __extends(Desktop, _super);
    function Desktop() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Desktop;
}(Computer));
var Laptop = /** @class */ (function (_super) {
    __extends(Laptop, _super);
    function Laptop() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    // Fields
    // Methods
    Laptop.prototype.fold = function () {
        console.log("-----------Folding Process ----------");
        console.log("Step1: Folding");
        console.log("");
    };
    return Laptop;
}(Computer));
var Walltop = /** @class */ (function (_super) {
    __extends(Walltop, _super);
    function Walltop() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return Walltop;
}(Computer));
// Objects
var computer;
// can be a Desktop Computer
computer = new Desktop("HP", "XP-X2", new Mouse(), new Intel(), new InternalMemory(), new Projector());
computer.setBrand("DELL");
console.log(computer.getBrand());
// can be a Laptop Computer
computer = new Laptop("HP", "XP-X2", new KeyBoard(), new AMD(), new SSD(), new Monitor());
computer.process('cccgdnbhx');
computer.setInputDevice(new Mouse());
console.log(computer.getInputDevice());
computer.input("");
// // can be a walltop Computer
computer = new Walltop("HP", "XP-X2", new KeyBoard(), new Nvidia, new HDD(), new Projector());
computer.store('cccgdnbhx');
computer.setStorageDevice(new SSD());
console.log(computer.getStorageDevice());
computer.store('x');
